model/src/main/kotlin/OrtResult.kt

```diff
@@ -94,6 +94,10 @@ data class OrtResult(
 
     private data class ProjectEntry(val project: Project, val isExcluded: Boolean)
 
+    /**
+     * A map of projects and their excluded state. Calculating this map once brings massive performance improvements
+     * when querying projects in large analyzer results.
+     */
```

Comment:
| This makes sense, because packages
which are not in the analyzer result cannot be excluded, and we also
could not provide a `PathExclude` or `ScopeExclude` which matches the
package.

If the code was using `!isExcluded` as synonym for `!isIncluded` I wouldn't fully agree with that. In particular I'm afraid of missing any assumptions like `!isExcluded(id) == true --> getPackage(id) != null`. Have you found a way to check that?

I believe the previous algorithm was not broken, but it was actually using `!isIncluded` as semantic for ``isExcluded`.
So: Will we have a bug now if a package is in the package list but referenced from any scope?

Reply:
> If the code was using `!isExcluded` as synonym for `isIncluded` I wouldn't fully agree with that. In particular I'm afraid of missing any assumptions like `!isExcluded(id) == true --> getPackage(id) != null`. Have you found a way to check that?

I checked all callers of the function and did not see one which uses it in this way. Also all callers of `getPackage()` do null handling, except for one in `AmazonOssAttributionBuilderReporter` which I will fix separately, because it was already wrong before this change. I also think that `!isExcluded` should not be used as synonym for `isIncluded`, instead the function should solely be used to check if a package is explicitly excluded. If it is called with an id that is not in the analyzer result at all the problem is likely on the caller side.

> I believe the previous algorithm was not broken, but it was actually using !isIncluded as semantic for ``isExcluded`.
So: Will we have a bug now if a package is in the package list but referenced from any scope?

It wasn't broken, but it was also not well documented, and the way it was implemented made it harder to fix the issues for handling projects and dependencies which are not in the package list. If a package is in the package list but not referenced from any scope it is impossible to exclude it, because it can only be excluded by excluding projects or scopes. In this case I would wonder why the package is there at all and assume a bug in a package manager implementation.

Reply:
> In this case I would wonder why the package is there at all and assume a bug in a package manager implementation.

I know that this case happens (not exactly where anymore), but I like the idea of fixing this in package manager.